### PR TITLE
Use internal LB's IP for intra-node communication

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- Add pre-kubeadm join commands to worker nodes
- Also pre-create a DNS Name for the frontend LB's IP that get assigned to the controlplane VM.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

- Below templates are _the only valid changes_
  - `templates/cluster-template-windows.yaml`
  - `templates/cluster-template.yaml`

- PR: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5210 should fix this issue. Below templates are _not expected to change_. Will probe further to update the kustomization.yaml for respective flavors
  - `templates/cluster-template-ephemeral.yaml`
  - `templates/cluster-template-edgezone.yaml`
  - `templates/cluster-template-azure-cni-v1.yaml`
  - `templates/cluster-template-azure-bastion.yaml`

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Use internal LB's IP for intra-node communication
```

### CODE REVIEW

File: templates/cluster-template.yaml
Comment (jackfrancis) [inline templates/cluster-template.yaml:36]: This is probably not going to be unique enough across all public DNS records for a given Azure region. We'll need an extra hash between "cluster name" and "location", probably.

File: hack/observability/opentelemetry/controller-manager-patch.yaml
Comment (nawazkh) [inline hack/observability/opentelemetry/controller-manager-patch.yaml:17]: Increases logging across CAPZ Controller manager. 
So `log.V(2).Info("haha")` while local dev works.

@willie-yao ☝🏼

File: azure/scope/cluster.go
Comment (jackfrancis) [inline azure/scope/cluster.go:268]: I think at this point the simplest, most correct condition for this is `if s.APIServerLB().Type != infrav1.Internal`. That's more consistent w/ the other changes, too.

File: azure/services/loadbalancers/loadbalancers.go
Comment (jackfrancis) [inline azure/services/loadbalancers/loadbalancers.go:71]: we can get rid of these two additions now

File: azure/services/networkinterfaces/spec.go
Comment (jackfrancis) [inline azure/services/networkinterfaces/spec.go:123]: nit: unnecessary newline

File: hack/observability/opentelemetry/controller-manager-patch.yaml
Comment (jackfrancis) [reply hack/observability/opentelemetry/controller-manager-patch.yaml:17]: we can remove this from the PR now

File: templates/cluster-template.yaml
Comment (willie-yao) [inline templates/cluster-template.yaml:199]: I don't want to over-complicate things, but would there ever be a reason for a user to want to customize the internal IP? If so, it might be worth having it be customizable with an environment variable. Otherwise it should be fine as-is.

File: templates/flavors/default/patches/control-plane.yaml
Comment (jackfrancis) [inline templates/flavors/default/patches/control-plane.yaml:8]: We can get rid of this comment, as (arguably) this is a normal configuration and not a hack, and the security rules referred to are not universal across Azure.

File: templates/cluster-template.yaml
Comment (jackfrancis) [inline templates/cluster-template.yaml:199]: It might be cleaner to do this in `postKubeadmCommands`:

```
- echo '10.0.0.100 $(grep 'server:' /etc/kubernetes/kubelet.conf | awk -F[/:] '{print $5}') >> /etc/hosts
```

This approach would allow us to entirely skip the need to pass in the dnsName manually in another part of the template, and we'd just rely upon the existing foo and kubeadm to tell us this info.

I'm pretty sure we'd need to do this as a `postKubeadmCommand`, and not `preKubeadmCommand`, kubeadm command. I would assume that kubeadm hasn't yet paved the kubeconfig by the time we invoke `preKubeadmCommands`.

File: azure/scope/cluster.go
Comment (jsturtevant) [inline azure/scope/cluster.go:283]: This could be made configurable via the spec in the future as @willie-yao mentioned below. Might be worth opening an issue for improvement

File: templates/cluster-template.yaml
Comment (jsturtevant) [reply templates/cluster-template.yaml:199]: There is that option in the SPEC for private lbs. This would be a good improvement, but for now I wouldn't block on it.

File: templates/flavors/default/patches/control-plane.yaml
Comment (jsturtevant) [reply templates/flavors/default/patches/control-plane.yaml:8]: Another solution would be to update the IP address that is passed to CAPI kubeadm.  The approach taken here has the advantage of not needing to have the management cluster peered to the worker node vnet. I guess scenarios where you wanted it fully private you could go with a private cluster and bolt on a public IP address with ASO, if desired, later once https://github.com/Azure/azure-service-operator/issues/4368 is resolved.

Anyways keeping some form of comment here is probably useful for our future selves.

Comment (jackfrancis) [reply templates/flavors/default/patches/control-plane.yaml:8]: I like the idea of a comment that clarifies any housekeeping to ensure that nodes are able to connect to the apiserver locally (whether or not there is a private or public apiserver endpoint), but I don't think we want to reference a "security policy", which is not general for Azure, and is subject to change.

File: azure/scope/machine.go
Comment (nawazkh) [inline azure/scope/machine.go:299]:
```suggestion
			if m.IsAPIServerPrivate() {
				spec.InternalLBName = m.APIServerLBName()
				spec.InternalLBAddressPoolName = m.APIServerLBPoolName()
			} else {
				spec.InternalLBName = m.APIServerLBName() + "-ilb"
				spec.InternalLBAddressPoolName = m.APIServerLBPoolName() + "-ilb"
				spec.PublicLBNATRuleName = m.Name()
				spec.PublicLBAddressPoolName = m.APIServerLBPoolName()
			}
		}
```

I think this needs to be updated as above.

Comment (nawazkh) [reply azure/scope/machine.go]: Ignore commit name, but addressed this in the commit 0b5d19c017c66e19a5dd9950a3c4adfd031d2d9a

**Comment by nawazkh** [comment]
Dropping all the tweaks that enabled running Ginkgo tests locally.

File: templates/cluster-template.yaml
Comment (nawazkh) [reply templates/cluster-template.yaml:199]: > This approach would allow us to entirely skip the need to pass in the dnsName manually in another part of the template, and we'd just rely upon the existing foo and kubeadm to tell us this info.

I like the idea. One less variable to manage.

However, isn't it better to setup custom DNS resolution for the API Server in `preKubeadmCommands` than in `postKubeadmCommands`? 
By using `preKubeadmCommand`, in this scenario, we can be sure that no Azure network policies will interfere in the `kubeadm join`step. 
Meaning, if Azure policies (blocking internet access on the VM) were to kick in immediately after a VM comes up, custom DNS resolution for the API Server to the internal IP should help `kubeadm join`.  

What do you say?

File: templates/flavors/default/patches/control-plane.yaml
Comment (nawazkh) [reply templates/flavors/default/patches/control-plane.yaml:8]: > Another solution would be to update the IP address that is passed to CAPI kubeadm

Is that IP Address present in `/etc/kubernetes/kubelet.conf` as referenced in Jack's proposal here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5209#discussion_r1821104012 ?

Comment (nawazkh) [reply templates/flavors/default/patches/control-plane.yaml]: > I like the idea of a comment that clarifies any housekeeping to ensure that nodes are able to connect to the apiserver locally (whether or not there is a private or public apiserver endpoint), but I don't think we want to reference a "security policy", which is not general for Azure, and is subject to change.

Addressed via 4996a360703222d42f30b2566deff68d68eb966d

File: templates/flavors/default/patches/control-plane.yaml
Comment (nojnhuh) [inline templates/flavors/default/patches/control-plane.yaml:14]: Could we incorporate this into the base template for all the other self-managed templates that aren't derived from the `default` flavor?

File: templates/flavors/default/patches/kubeadm-config-template.yaml
Comment (nojnhuh) [inline templates/flavors/default/patches/kubeadm-config-template.yaml:4]: Can we add something similar for the `machinepool` flavor and for any other worker nodes that don't come from the `default` flavor?

File: api/v1beta1/tags.go
Comment (nojnhuh) [inline api/v1beta1/tags.go:124]: Is the internal load balancer required for the /etc/hosts change to work? Or is this orthogonal?

Comment (nawazkh) [reply api/v1beta1/tags.go:124]: Yes, internal load balancer is required for the /etc/hosts fix to work. 
The IP address of the load-balancer (defaulted to 10.0.0.100) is used to resolve APIServer's public DNS name via the update in `/etc/hosts`.

File: templates/flavors/default/patches/control-plane.yaml
Comment (nawazkh) [reply templates/flavors/default/patches/control-plane.yaml:14]: Waiting on a resolution here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5209#discussion_r1821104012, and then will pull this change into `base`.

File: templates/cluster-template.yaml
Comment (jackfrancis) [reply templates/cluster-template.yaml:199]: Will `/etc/kubernetes/kubelet.conf` be present at the time that `preKubeadmCommands` runs? I assume that kubeadm is responsible for paving that file, which means we have to wait.

But that may not be practical for Windows nodes that come online later (it may be too late and as you mention, kubeadm join may fail and block `postKubeadmCommands` from ever executing.

cc @jsturtevant

File: templates/flavors/default/patches/control-plane.yaml
Comment (jackfrancis) [reply templates/flavors/default/patches/control-plane.yaml:14]: We can probably land these changes to *all* templates in a phased approach. Landing a working pattern of this ASAP will allow folks managing their own templates to integrate with this new solution sooner.

wdyt @nojnhuh?

Comment (nojnhuh) [reply templates/flavors/default/patches/control-plane.yaml:14]: sgtm

Comment (nawazkh) [reply templates/flavors/default/patches/control-plane.yaml:14]: Actually, why don't I follow it up with another PR and unblock the team asap with landing this one ?

File: templates/flavors/default/patches/kubeadm-config-template.yaml
Comment (nawazkh) [reply templates/flavors/default/patches/kubeadm-config-template.yaml:4]: I will follow up with another PR.

**Review by jackfrancis — APPROVED** [review]
/lgtm
/approve

File: templates/cluster-template.yaml
Comment (jsturtevant) [reply templates/cluster-template.yaml:199]: > Will /etc/kubernetes/kubelet.conf be present at the time that preKubeadmCommands runs?

I don't think it will be present, kubeadm creates it.

**Comment by jsturtevant** [comment]
/lgtm

Thanks for helping fix this up!

**Comment by nojnhuh** [comment]
Can we drop the hold now that commits are squashed?

**Pull request merged**
